packages/generator-helper/src/GeneratorProcess.ts

```diff
@@ -20,192 +21,215 @@ type GeneratorProcessOptions = {
 }
 
 export class GeneratorError extends Error {
-  public code: number
-  public data?: any
-  constructor(message: string, code: number, data?: any) {
+  name = 'GeneratorError'
+
+  constructor(message: string, public code?: number, public data?: any) {
     super(message)
-    this.code = code
-    this.data = data
     if (data?.stack) {
       this.stack = data.stack
     }
   }
 }
 
+type ResultHandler<T = unknown> = {
+  resolve: (value: T) => void
+  reject: (error: Error) => void
+}
+
 export class GeneratorProcess {
-  child?: ChildProcessByStdio<any, any, any>
-  listeners: { [key: string]: (result: any, err?: Error) => void } = {}
-  private stderrLogs = ''
+  private child?: ChildProcessByStdio<Writable, null, Readable>
+  private handlers: Record<string, ResultHandler> = {}
   private initPromise?: Promise<void>
   private isNode: boolean
-  private initWaitTime: number
-  private currentGenerateDeferred?: {
-    resolve: (result: any) => void
-    reject: (error: Error) => void
-  }
-  constructor(private executablePath: string, { isNode = false, initWaitTime = 200 }: GeneratorProcessOptions = {}) {
+  private errorLogs = ''
+  private pendingError: Error | undefined
+
+  constructor(private pathOrCommand: string, { isNode = false }: GeneratorProcessOptions = {}) {
     this.isNode = isNode
-    this.initWaitTime = initWaitTime
   }
+
   async init(): Promise<void> {
     if (!this.initPromise) {
       this.initPromise = this.initSingleton()
     }
     return this.initPromise
   }
+
   initSingleton(): Promise<void> {
     return new Promise((resolve, reject) => {
-      try {
-        if (this.isNode) {
-          this.child = fork(this.executablePath, [], {
-            stdio: ['pipe', 'inherit', 'pipe', 'ipc'],
-            env: {
-              ...process.env,
-              PRISMA_GENERATOR_INVOCATION: 'true',
-            },
-            execArgv: ['--max-old-space-size=8096'],
-          })
-        } else {
-          this.child = spawn(this.executablePath, {
-            stdio: ['pipe', 'inherit', 'pipe'],
-            env: {
-              ...process.env,
-              PRISMA_GENERATOR_INVOCATION: 'true',
-            },
-            shell: true,
-          })
+      if (this.isNode) {
+        this.child = fork(this.pathOrCommand, [], {
+          stdio: ['pipe', 'inherit', 'pipe', 'ipc'],
```

Comment:
Why is `ipc` needed here?

Reply:
It was already here, I didn't change it. What it does is it creates an IPC channel for passing messages or transferring file descriptors between parent and child, i.e., enables `child.send()` in the parent and `process.on('message')` and `process.send()` in the child. I don't think we use it, so we could try removing this, but I'd probably do that in another PR since it's a bit out of scope.